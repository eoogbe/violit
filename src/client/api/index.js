/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

export { load, loadAll } from './load';
export { default as InvalidResponseError } from './InvalidResponseError';
export { default as checkExists } from './checkExists';
export { default as checkPermissions } from './checkPermissions';
export { default as href } from './href';
export { default as sendCreate } from './sendCreate';
export { default as sendDelete } from './sendDelete';
export { default as sendPost } from './sendPost';
