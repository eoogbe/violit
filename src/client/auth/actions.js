/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import createActions from 'client/utils/createActions';

export default createActions('auth', [
  'LOAD_TRIGGERED',
  ['LOAD_SUCCEEDED', 'user'],
  'LOAD_FAILED',
  ['LOGIN_TRIGGERED', 'credentials'],
  ['LOGIN_SUCCEEDED', 'user'],
  'LOGIN_FAILED',
  ['CREDENTIALS_INVALID', 'error'],
  'LOGOUT_TRIGGERED',
  ['SIGNUP_TRIGGERED', 'user'],
  ['SIGNUP_SUCCEEDED', 'user'],
  ['SIGNUP_FAILED'],
  ['CHECK_HAS_USER_TRIGGERED', 'username'],
  ['CHECK_HAS_USER_SUCCEEDED', 'hasUser'],
  'CHECK_HAS_USER_FAILED',
]);
