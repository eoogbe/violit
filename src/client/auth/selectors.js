/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import { createSelector } from 'reselect';
import { NAME } from './constants';

/** Selects the auth state. */
export const getAuth = R.prop(NAME);

/** Selects the credentials of the current user. */
export const getCredentials = createSelector(
  [getAuth],
  R.prop('credentials')
);

/** Selects the username of the current user. */
export const getUsername = createSelector(
  [getAuth],
  R.path(['credentials', 'username'])
);
