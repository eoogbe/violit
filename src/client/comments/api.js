/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import {
  checkPermissions,
  href,
  loadAll,
  sendCreate,
  sendPost,
} from 'client/api';

/** Converts a comment from the API format to the view format. */
const convertComment = ({ links, ...comment }) => R.compose(
  R.assoc('canDelete', !!links.deleteAction),
  R.pick(['id', 'dateCreated', 'dateDeleted', 'text', 'threadSlug', 'author'])
)(comment);

/** Loads comments from the server. */
export const loadComments = async (threadSlug) => {
  const { data } = await loadAll(href(`/threads/${threadSlug}/comments`));
  return data.map(convertComment);
};

/**
 * Sends the specified `comment` to the server to be created.
 */
export const sendCreateComment = async (comment) => {
  const url = href(`/threads/${comment.threadSlug}/comments`);
  const { body } = await sendCreate(url, ['id', 'text'], comment);
  return convertComment(body);
};

/** Checks if the current user can create a comment. */
export const checkCanCreateComment = async (threadSlug) => {
  const url = href(`/threads/${threadSlug}/comments`);
  const methods = await checkPermissions(url);
  return methods.includes('POST');
};

/** Sends a comment to be deleted on the server. */
export const sendDeleteComment = async (commentId) => {
  await sendPost(href(`/comments/${commentId}/delete`));
  return null;
};
