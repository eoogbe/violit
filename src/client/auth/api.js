/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as R from 'ramda';
import { checkExists, href, load, sendCreate, sendDelete } from 'client/api';

/** Creates the specified `credentials` on the server. */
const createCredentials = async (url, credentials, {
  status,
  errorMessage,
 }) => {
  try {
    const { body } = await sendCreate(
      url,
      ['username', 'password'],
      credentials
    );
    return R.pick(['username'], body);
  } catch (err) {
    if (R.path(['response', 'status'], err) !== status) {
      throw new Error(errorMessage);
    }

    const message = R.path(['response', 'body', 'error', 'message'], err);
    if (!message) {
      throw new Error(errorMessage);
    }

    return { error: message };
  }
};

/** Loads the current user's credentials from the server if they exist. */
export const loadCredentials = async () => {
  try {
    const credentials = await load(href('/auth'));
    return R.pick(['username'], credentials);
  } catch (err) {
    if (R.path(['response', 'status'], err) !== 405) {
      throw new Error('Something went wrong with loading the credentials.');
    }

    return null;
  }
};

/** Authenticates a user. */
export const login = (credentials) =>
  createCredentials(href('/auth'), credentials, {
    status: 404,
    errorMessage: 'Something went wrong with logging in.'
  });

/** Logs a user out of the current session. */
export const logout = async () => {
  await sendDelete(href('/auth'));
  return null;
};

/** Registers the specified `user` on the server. */
export const signup = (user) =>
  createCredentials(href('/users'), user, {
    status: 422,
    errorMessage: 'Something went wrong with signing up.'
  });

/** Checks if a user with the specified `username` exists on the server. */
export const checkHasUser = (username) =>
  checkExists(href(`/users/${username}`));
