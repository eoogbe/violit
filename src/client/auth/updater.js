/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  concat,
  constantState,
  decorate,
  handleAction,
  updateStateAt,
  withDefaultState,
} from 'redux-fp';
import matchActions from 'client/utils/matchActions';
import actions from './actions';

const INITIAL_STATE = {};

/** Updater that sets the current user's credentials. */
const setCredentials =
  decorate(
    matchActions([
      actions.loadSucceeded.type,
      actions.loginSucceeded.type,
      actions.signupSucceeded.type,
    ]),
    updateStateAt('credentials'),
    ({ payload }) => () => payload.user
  );

/** Updater that unsets the current user's credentials. */
const unsetCredentials =
  decorate(
    handleAction(actions.logoutTriggered.type),
    updateStateAt('credentials'),
    constantState(null)
  );

/** The auth updater. */
export default decorate(
  withDefaultState(INITIAL_STATE),
  concat(
    setCredentials,
    unsetCredentials,
  )
);
