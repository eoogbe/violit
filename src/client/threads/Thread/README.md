```jsx
import { MemoryRouter, Route } from 'react-router-dom';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';
import { constants as authConstants } from 'client/auth';

mockFetch([
  THREADS.threadFetch1,
  COMMENTS.canCreateCommentFetch,
  COMMENTS.commentsFetch,
]);

const state = {
  [authConstants.NAME]: {
    credentials: USER.credentials,
  },
};

<StoreProvider state={state}>
  <MemoryRouter initialEntries={[THREADS.thread1.path]}>
    <Route path="/threads/:slug">
      <Thread />
    </Route>
  </MemoryRouter>
</StoreProvider>
```
