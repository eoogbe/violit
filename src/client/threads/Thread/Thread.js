/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, { useRef, useState } from 'react';
import { useParams } from 'react-router-dom';
import styled from 'styled-components';
import { useCredentials } from 'client/auth';
import { CommentList, api as commentsApi } from 'client/comments';
import { Alert } from 'client/core';
import { LoadingContainer } from 'client/loading';
import ThreadMain from '../ThreadMain';
import * as api from '../api';
import useOptimisticPrepend from '../useOptimisticPrepend';

const StyledThread = styled.article`
  font-size: 1rem;
  padding: ${({ theme }) => theme.spacingMd};
  position: relative;
`;

const StyledMainContainer = styled.div`
  border-bottom: 1px solid ${({ theme }) => theme.gray200};
`;

/** The screen for a thread of discussion. */
const Thread = () => {
  const { slug } = useParams();
  const credentials = useCredentials();
  const [deleteError, setDeleteError] = useState(false);

  const commentListEl = useRef();
  const [
    createdComments,
    createError,
    createComment,
  ] = useOptimisticPrepend((commentId, { text }) => ({
    id: commentId,
    author: credentials.username,
    dateCreated: new Date().toISOString(),
    text: text.trimEnd(),
    threadSlug: slug,
  }), commentsApi.sendCreateComment, {
    onPrepend: () => {
      commentListEl.current.scrollIntoView();
    },
  });

  return (
    <StyledThread
      itemScope
      itemType="http://schema.org/DiscussionForumPosting"
    >
      <StyledMainContainer>
        <LoadingContainer
          loadData={() => api.loadThread(slug)}
          action="loading the thread"
          render={(thread) => (
            <ThreadMain
              loggedIn={!!credentials}
              createComment={createComment}
              onDeleteError={() => setDeleteError(true)}
              {...thread}
            />
          )}
        />
      </StyledMainContainer>
      <CommentList
        threadSlug={slug}
        additionalComments={createdComments}
        ref={commentListEl}
      />
      {createError && <Alert action="creating a comment" />}
      {deleteError && <Alert action="deleting the thread" />}
    </StyledThread>
  );
};

export default Thread;
