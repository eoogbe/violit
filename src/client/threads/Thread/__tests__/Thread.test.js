/**
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React from 'react';
import { mount } from 'enzyme';
import { act } from 'react-dom/test-utils';
import { Provider } from 'react-redux';
import { MemoryRouter, Route } from 'react-router-dom';
import TestRenderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import * as COMMENTS from '__fixtures__/commentsFixture';
import * as THREADS from '__fixtures__/threadsFixture';
import * as USER from '__fixtures__/userFixture';
import { constants as authConstants } from 'client/auth';
import { CommentForm } from 'client/comments';
import { Alert } from 'client/core';
import { Input } from 'client/forms';
import mockFetch from 'testUtils/mockFetch';
import { updateWrapper, wait } from 'testUtils/wait';
import Thread from '../Thread';

const NOW = 1572566400000;

const mockStore = configureStore();

const expectRender = async (responses, { loggedIn }) => {
  Date.now = jest.fn().mockReturnValue(NOW);
  mockFetch(responses);

  const state = loggedIn && {
    [authConstants.NAME]: {
      credentials: USER.credentials,
    },
  };
  const store = mockStore(state);

  let renderer;
  await TestRenderer.act(async () => {
    renderer = TestRenderer.create(
      <Provider store={store}>
        <MemoryRouter initialEntries={[THREADS.thread1.path]}>
          <Route path="/threads/:slug">
            <Thread />
          </Route>
        </MemoryRouter>
      </Provider>
    );

    await wait();
  });

  return expect(renderer.toJSON()).toMatchSnapshot();
};

const simulateCommentFormSubmit = (wrapper) => {
  const onSubmit = wrapper.find(CommentForm).children().prop('onSubmit');
  return act(async () => {
    await onSubmit({
      preventDefault: () => {},
    });
    await updateWrapper(wrapper);
  });
};

it('renders without crashing', () => {
  return expectRender([
    THREADS.threadFetch1,
    COMMENTS.canCreateCommentFetch,
    COMMENTS.commentsFetch,
  ], { loggedIn: true });
});

// TOOD(#68): Figure out how to manipulate an enzyme wrapper for an element
// loaded asynchronously
it.skip('when submitting comment form scrolls to comment list', async () => {
  mockFetch([
    THREADS.threadFetch1,
    COMMENTS.canCreateCommentFetch,
    COMMENTS.commentsFetch,
  ]);

  const scrollIntoView = jest.fn();
  window.HTMLElement.prototype.scrollIntoView = scrollIntoView;

  const store = mockStore({
    [authConstants.NAME]: {
      data: USER.credentials,
    },
  });

  const wrapper = mount(
    <Provider store={store}>
      <MemoryRouter initialEntries={[THREADS.thread1.path]}>
        <Route path="/threads/:slug">
          <Thread />
        </Route>
      </MemoryRouter>
    </Provider>
  );

  await wait();

  wrapper.find(Input).simulate('change', {
    target: { name: 'text', value: 'Lorem ipsum' },
  });

  await simulateCommentFormSubmit(wrapper);

  expect(scrollIntoView).toHaveBeenCalled();
});

// TOOD(#68): Figure out how to manipulate an enzyme wrapper for an element
// loaded asynchronously
it.skip('when delete fails shows an alert', async () => {
  mockFetch([
    {
      url: THREADS.thread1.deleteUrl,
      method: 'POST',
      status: 500,
    },
  ]);

  const store = mockStore({
    [authConstants.NAME]: {
      data: USER.credentials,
    },
  });

  let wrapper;
  await act(async () => {
    wrapper = mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[THREADS.thread1.path]}>
          <Route path="/threads/:slug">
            <Thread />
          </Route>
        </MemoryRouter>
      </Provider>
    );

    await wait();
  });

  expect(wrapper.find(Alert)).toHaveLength(0);

  await act(async () => {
    wrapper.find('OverflowToggle').simulate('click');
    wrapper.find('MenuItemButton').first().simulate('click');

    await wait();
  });

  expect(wrapper.find(Alert)).toHaveLength(1);
});
