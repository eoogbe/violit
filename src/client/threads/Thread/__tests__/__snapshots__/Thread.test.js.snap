// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<article
  className="sc-jDwBTQ gWsmob"
  itemScope={true}
  itemType="http://schema.org/DiscussionForumPosting"
>
  <div
    className="sc-gPEVay dOeLqP"
  >
    <h2
      className="sc-brqgnP kAzWxO"
      itemProp="headline"
    >
      Lorem ipsum
    </h2>
    <header
      className="sc-jKJlTe dCQNcG"
    >
      <span
        className="sc-eNQAEJ jsbjCq"
        itemProp="author"
      >
        fred-astaire
      </span>
      <time
        dateTime="2019-09-29T01:23:45.679Z"
        itemProp="dateCreated"
        title="2019-09-29T01:23:45.679Z"
      >
        1 month ago
      </time>
    </header>
    <div
      itemProp="articleBody"
    >
      <div
        className="sc-bZQynM kESxfG"
      >
        <p>
          Dolor sit amet, consectetur adipiscing elit.
        </p>
      </div>
    </div>
    <form
      className="sc-cSHVUG jxHEP"
      onSubmit={[Function]}
    >
      <ul
        className="sc-kAzzGY ckrMOY"
      >
        <li>
          <textarea
            className="sc-htoDjs kEmxlR"
            name="text"
            onChange={[Function]}
            placeholder="Write a comment"
            rows={5}
            value=""
          />
        </li>
      </ul>
      <button
        className="sc-ifAKCX hbBAat sc-chPdSV bBcFRQ"
        disabled={true}
        type="submit"
      >
        Comment
      </button>
    </form>
    <footer
      className="sc-kpOJdX fzEkOA"
    >
      <nav
        className="sc-dxgOiQ qAjpS"
      >
        <button
          aria-controls="overflow-menu-thread-thread1"
          aria-expanded={false}
          aria-haspopup="true"
          className="sc-ckVGcZ ehJtkF"
          onClick={[Function]}
          type="button"
        >
          <svg
            fill="currentColor"
            focusable={false}
            height="1.5em"
            size="1.5em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            title="More actions"
            viewBox="0 0 24 24"
            width="1.5em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
            />
          </svg>
        </button>
      </nav>
    </footer>
  </div>
  <ul
    className="sc-cvbbAY HtGJA"
  >
    <li
      className="sc-iAyFgw eoNjzy"
      itemProp="comment"
      itemScope={true}
      itemType="http://schema.org/Comment"
    >
      <article>
        <header
          className="sc-jKJlTe dCQNcG"
        >
          <span
            className="sc-eNQAEJ jsbjCq"
            itemProp="author"
          >
            fred-astaire
          </span>
          <time
            dateTime="2019-10-09T10:11:12.123Z"
            itemProp="dateCreated"
            title="2019-10-09T10:11:12.123Z"
          >
            3 weeks ago
          </time>
        </header>
        <div
          itemProp="text"
        >
          <div
            className="sc-bZQynM kESxfG"
          >
            <p>
              Sed do eiusmod tempor incididunt ut labore et dolore.
            </p>
          </div>
        </div>
        <footer
          className="sc-kpOJdX fzEkOA"
        >
          <nav
            className="sc-dxgOiQ qAjpS"
          >
            <button
              aria-controls="overflow-menu-comment-comment1"
              aria-expanded={false}
              aria-haspopup="true"
              className="sc-ckVGcZ ehJtkF"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="currentColor"
                focusable={false}
                height="1.5em"
                size="1.5em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                title="More actions"
                viewBox="0 0 24 24"
                width="1.5em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </button>
          </nav>
        </footer>
      </article>
    </li>
    <li
      className="sc-iAyFgw eoNjzy"
      itemProp="comment"
      itemScope={true}
      itemType="http://schema.org/Comment"
    >
      <article>
        <header
          className="sc-jKJlTe dCQNcG"
        >
          <span
            className="sc-eNQAEJ jsbjCq"
            itemProp="author"
          >
            fred-astaire
          </span>
          <time
            dateTime="2019-10-08T07:08:09.101Z"
            itemProp="dateCreated"
            title="2019-10-08T07:08:09.101Z"
          >
            3 weeks ago
          </time>
        </header>
        <div
          itemProp="text"
        >
          <p
            className="sc-hSdWYo kGFdpB"
          >
            [deleted]
          </p>
        </div>
      </article>
    </li>
    <li
      className="sc-iAyFgw eoNjzy"
      itemProp="comment"
      itemScope={true}
      itemType="http://schema.org/Comment"
    >
      <article>
        <header
          className="sc-jKJlTe dCQNcG"
        >
          <span
            className="sc-eNQAEJ jsbjCq"
            itemProp="author"
          >
            fred-astaire
          </span>
          <time
            dateTime="2019-09-30T01:02:03.456Z"
            itemProp="dateCreated"
            title="2019-09-30T01:02:03.456Z"
          >
            1 month ago
          </time>
        </header>
        <div
          itemProp="text"
        >
          <div
            className="sc-bZQynM kESxfG"
          >
            <p>
              Quis nostrud exercitation ullamco laboris nisi ut.
            </p>
          </div>
        </div>
        <footer
          className="sc-kpOJdX fzEkOA"
        >
          <nav
            className="sc-dxgOiQ qAjpS"
          >
            <button
              aria-controls="overflow-menu-comment-comment3"
              aria-expanded={false}
              aria-haspopup="true"
              className="sc-ckVGcZ ehJtkF"
              onClick={[Function]}
              type="button"
            >
              <svg
                fill="currentColor"
                focusable={false}
                height="1.5em"
                size="1.5em"
                stroke="currentColor"
                strokeWidth="0"
                style={
                  Object {
                    "color": undefined,
                  }
                }
                title="More actions"
                viewBox="0 0 24 24"
                width="1.5em"
                xmlns="http://www.w3.org/2000/svg"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </button>
          </nav>
        </footer>
      </article>
    </li>
  </ul>
</article>
`;
